pkg/controller/druid/handler.go

```diff
@@ -390,6 +394,151 @@ func checkCrashStatus(sdk client.Client, m *v1alpha1.Druid) {
+
+	pvcList := makePersistentVolumeClaimListEmptyObj()
+	listOpts := []client.ListOption{
+		client.InNamespace(drd.Namespace),
+		client.MatchingLabels(map[string]string{
+			"component": nodeSpec.NodeType,
+		}),
+	}
+
+	if err := sdk.List(context.TODO(), pvcList, listOpts...); err != nil {
+		e := fmt.Errorf("failed to list pvc for [%s:%s] due to [%s]", drd.Kind, drd.Name, err.Error())
+		sendEvent(sdk, drd, v1.EventTypeWarning, "LIST_FAIL", e.Error())
+		logger.Error(e, e.Error(), "name", drd.Name, "namespace", drd.Namespace)
+		return nil, e
+	}
+
+	// create a slice []*pvc
+	pvc := make([]*v1.PersistentVolumeClaim, 0)
+	for i := range pvcList.Items {
+		pvc = append(pvc, &pvcList.Items[i])
+	}
+
+	return pvc, nil
+
+}
+
+// desVolumeClaimTemplateSize: the druid CR holds this value for a sts volumeclaimtemplate
+// currVolumeClaimTemplateSize: the sts owned by druid CR holds this value in volumeclaimtemplate
+// pvcSize: the pvc referenced by the sts holds this value
+// type of vars is resource.Quantity. ref: https://godoc.org/k8s.io/apimachinery/pkg/api/resource
+func getVolumeClaimTemplateSizes(sts *appsv1.StatefulSet, nodeSpec *v1alpha1.DruidNodeSpec, pvc []*v1.PersistentVolumeClaim) (desVolumeClaimTemplateSize, currVolumeClaimTemplateSize, pvcSize resource.Quantity) {
+
+	for i := range nodeSpec.VolumeClaimTemplates {
+		desVolumeClaimTemplateSize = nodeSpec.VolumeClaimTemplates[i].Spec.Resources.Requests[v1.ResourceStorage]
+	}
+
+	for i := range sts.Spec.VolumeClaimTemplates {
+		currVolumeClaimTemplateSize = sts.Spec.VolumeClaimTemplates[i].Spec.Resources.Requests[v1.ResourceStorage]
+	}
+
+	for i := range pvc {
+		pvcSize = pvc[i].Spec.Resources.Requests[v1.ResourceStorage]
+	}
+
+	return desVolumeClaimTemplateSize, currVolumeClaimTemplateSize, pvcSize
+
+}
+
+func checkIfStsHasEverythingElseSame(sdk client.Client, drd *v1alpha1.Druid, sts *appsv1.StatefulSet, nodeSpec *v1alpha1.DruidNodeSpec) bool {
+
+	if nodeSpec.PodManagementPolicy != sts.Spec.PodManagementPolicy {
+		e := fmt.Errorf("StatefulSet [%s] in namespace is invalid [%s],  Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden. ", sts.Name, sts.Namespace)
+		sendEvent(sdk, drd, v1.EventTypeWarning, "UPDATE_FAIL", e.Error())
+		logger.Error(e, e.Error(), "name", drd.Name, "namespace", drd.Namespace)
+		return false
+	}
+
+	/*
+		annotations := reflect.DeepEqual(nodeSpec.PodAnnotations, sts.Annotations)
+		if !annotations {
+			e := fmt.Errorf("StatefulSet [%s] in namespace is invalid [%s],  Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden. ", sts.Name, sts.Namespace)
+			sendEvent(sdk, drd, v1.EventTypeWarning, "UPDATE_FAIL", e.Error())
+			logger.Error(e, e.Error(), "name", drd.Name, "namespace", drd.Namespace)
+			return false
+		}
+	*/
+
+	return true
+}
+
```

Comment:
@himanshug i have a few question here, when we want to make sure everythingelseissame(), that means
1. metadata should be same ( labels + annotations )
2. podmanagementpolicy should be same
3. Each field inside volumeclaimtemplates should be same
The operator adds labels such as component etc, in that case nodeSpec.Podlables != sts.Labels , this condition will always be true......
any better approach/simpler approach to check, else the only options seems to check each field ?

Reply:
@himanshug do have a look at this query, kind of confused for this function. The rest functions have been tested out.

Reply:
haven't looked in detail, but would something like following work?

say, you have `oldSts` and `newSts` ... and you wanted to check everything else except `VolumeClaimTemplates` is same ... then do

```
oldSts.Spec.VolumeClaimTemplates = newSts.Spec.VolumeClaimTemplates

is reflect.DeepEquals(oldSts, newSts) ?
```